### PR TITLE
NEW: CLI For TileServer

### PR DESCRIPTION
**Pull request by measty**
Changed this commit to just add a CLI for setting up a tileserver

### CODE REVIEW

File: tiatoolbox/cli/overlay_from_json.py
Comment (shaneahmed) [inline tiatoolbox/cli/overlay_from_json.py:30]: Can we not avoid pyvips dependancy? 
Adding it would mean further restrictions on dependency versions.

Comment (John-P) [reply tiatoolbox/cli/overlay_from_json.py:30]: Indeed tifffile / zarr cna be used instead here.

**Comment by John-P** [comment]
I like this idea. However, my main concern is with the code structure. It doesn't make sense to have the whole implementation in a cli function. I would implement rendering if tiles from an annotation store first. Then a function to write all the tiles to a tiff with tiffile/zarr. Then make a wrapper cli argument which loads the geojson (or ndgeojson) to a store, then renders is out.

File: tests/test_tileserver.py
Comment (shaneahmed) [inline tests/test_tileserver.py:29]: why do you need this?

File: tiatoolbox/cli/tile_server.py
Comment (shaneahmed) [inline tiatoolbox/cli/tile_server.py:15]: Can we not run the server in a directory and open image directly from there?

**Review by shaneahmed — COMMENTED** [review]
Thanks Mark. Please can you address the comments.

File: tests/test_tileserver.py
Comment (measty) [reply tests/test_tileserver.py:29]: Hi Shan,
I use it in line 40: `np.zeros(wsi.slide_dimensions(1.25, "power"), dtype=np.uint8).T` to make an array of suitable dimensions to test the case where a low-res heatmap is given as an np array for one of the layers

File: tiatoolbox/cli/tile_server.py
Comment (measty) [reply tiatoolbox/cli/tile_server.py:15]: do you mean just give a directory and try to open all images in there as layers?

Comment (shaneahmed) [reply tiatoolbox/cli/tile_server.py:15]: We can open individual images here, but yes we should be able to open a directory in future.

Comment (shaneahmed) [reply tiatoolbox/cli/tile_server.py:15]: I am resolving it for the sake of this PR.

**Review by shaneahmed — APPROVED** [review]
Thanks @measty 
I will wait for @John-P's input before I merge it.

File: tiatoolbox/cli/tile_server.py
Comment (shaneahmed) [inline tiatoolbox/cli/tile_server.py:10]: shouldn't this be using decorator from `cli/common.py` for image input? This will automatically add help for that function as well. https://github.com/TissueImageAnalytics/tiatoolbox/blob/446f2e1e05f79d4e397267a2dea4f962c506fdd1/tiatoolbox/cli/common.py#L30

Comment (measty) [reply tiatoolbox/cli/tile_server.py:10]: the img_input decorator will only allow a single path (to a file or directory). But the tileserver cli needs to be able to accept a variable number of paths.
for example if i had a slide and heatmaps from two different models i need to be able to do:

tiatoolbox tileserver path/to/slide path/to/heatmap1 path/to/heatmap2

to start the tileserver with the slide and two additional layers for the heatmaps

Comment (shaneahmed) [reply tiatoolbox/cli/tile_server.py:10]: You can still use `img_input` by defining defaults and options. See for example, https://github.com/TissueImageAnalytics/tiatoolbox/blob/c7fe7851cf78ab582e8e02bc1b21d5227063fc77/tiatoolbox/cli/common.py#L50 You can see different file types have been called here https://github.com/TissueImageAnalytics/tiatoolbox/blob/c7fe7851cf78ab582e8e02bc1b21d5227063fc77/tiatoolbox/cli/patch_predictor.py#L32 
Similarly, options for `mode` have been modified https://github.com/TissueImageAnalytics/tiatoolbox/blob/c7fe7851cf78ab582e8e02bc1b21d5227063fc77/tiatoolbox/cli/patch_predictor.py#L35

Comment (measty) [reply tiatoolbox/cli/tile_server.py:10]: the only options available for img_input is the help option. 

https://github.com/TissueImageAnalytics/tiatoolbox/blob/c7fe7851cf78ab582e8e02bc1b21d5227063fc77/tiatoolbox/cli/common.py#L30-L32

I could change that to allow more options to cover the useage needed here, if that would be a better way to do it?

Comment (shaneahmed) [reply tiatoolbox/cli/tile_server.py:10]: yes that's fine you can modify it by keeping the previous options as default options.

File: tiatoolbox/visualization/tileserver.py
Comment (measty) [reply tiatoolbox/visualization/tileserver.py:42]: >I did this so that it isn't necessary to give names to the layers. You can give it a list and have autogenerated layer names ('layer1', 'layer2', etc), or a dict which specifies names for each layer via the key.

Although, thinking on that, there isn't a provision for specifying the layer names in the cli command, and there probably should be. I can add that in later today.

Will change the hint as suggested also.

File: tiatoolbox/cli/overlay_from_json.py
Comment (John-P) [reply tiatoolbox/cli/overlay_from_json.py]: I think this is resolved now.

File: tests/test_tileserver.py
Comment (John-P) [inline tests/test_tileserver.py:28]: Any particular reason why you are creating a Cli handler and then calling that rather than just using the slide_thumbnail method on WSIReader? Just seems like an odd way to make a thumbnail for the test.

File: tiatoolbox/visualization/tileserver.py
Comment (John-P) [inline tiatoolbox/visualization/tileserver.py:61]: @measty This could be better written as just `enumerate(layers)` as this will enumerate the keys. This should have been flagged with pre-commit. Are you running pre-commit before committing?

File: tiatoolbox/visualization/tileserver.py
Comment (John-P) [inline tiatoolbox/visualization/tileserver.py:65]: p is not a good variable name. I think this should have been flagged by pre-commit. I suggest something descriptive like `layer_path` instead.

File: tiatoolbox/visualization/tileserver.py
Comment (John-P) [inline tiatoolbox/visualization/tileserver.py:76]: It would be nice to be able to change the colormap. I'm personally not a big fan of jet as a hardcoded colormap because it is not perceptually uniform. The luminance does not correspond to the value, hence it is not good for colour blind users. If we must have a hardcoded map or a default I suggest "viridis" (the matplotlib default) or "magma". See here for more information:
- https://matplotlib.org/3.5.0/tutorials/colors/colormaps.html
- https://jakevdp.github.io/blog/2014/10/16/how-bad-is-your-colormap/

File: tiatoolbox/visualization/tileserver.py
Comment (John-P) [inline tiatoolbox/visualization/tileserver.py:80]: This would section could possibly be refactored into a function e.g. `colourise_image` which takes an ndarray and applies a cmap if is is a single channel array, else returns the image unmodified. Then this could be wrapped in VirtualWSIReader. E.g.

```python
if isinstance(layer, (str, Path)):
    p = Path(layer)
      if p.suffix in [".jpg", ".png"]:  # maybe add .webp and .gif here?
          # assume its a low-res heatmap
            layer = Image.open(p)
            layer = np.array(layer)
            layer = colourised_layer(layer, cmap=...)
            layer = VirtualWSIReader(layer, info=meta)
      else:
           layer = WSIReader.open(p)
```

I think this would make it a bit more readable / maintainable. This also does not account for low res TIFFs which we may need to address later.

**Review by John-P — CHANGES_REQUESTED** [review]
Looks good. Just a few suggestions for changes which might improve the code quality / maintainability in my opinion.

File: tiatoolbox/cli/tile_server.py
Comment (John-P) [reply tiatoolbox/cli/tile_server.py]: Is this resolved?

**Comment by John-P** [comment]
Just a side note, can we be consistent with comments capitalisation. I think that we should consistently write comments as a full sentence with a capital letter and full stop unless it is just a word or two note. I recently made a PR to tidy inconsistencies such as this up but I see them creeping in again.

File: tiatoolbox/visualization/tileserver.py
Comment (measty) [reply tiatoolbox/visualization/tileserver.py]: changed the default to viridis, and can choose the colourmap in colourise function, though there is no provision for choosing colourmap in cli at the moment. Also addressed the other code maintainablility/readability issues above

File: tests/test_tileserver.py
Comment (measty) [reply tests/test_tileserver.py]: i think it was a historical artifact of something i tried and removed, cant quite remember now! but have changed it

File: tiatoolbox/visualization/tileserver.py
Comment (measty) [reply tiatoolbox/visualization/tileserver.py]: moved the colourising code into a function in utils/visualisation. Haven't added any additional formats for now, can maybe look at that later

File: tiatoolbox/utils/visualization.py
Comment (John-P) [inline tiatoolbox/utils/visualization.py:40]: Just minor formatting. The first line of the docstring should be a single line with a full stop.

Additional detail is then in a paragraph after a blank line. Also missing args and returns here.

File: tiatoolbox/cli/tile_server.py
Comment (John-P) [inline tiatoolbox/cli/tile_server.py:11]: This docstring formatting needs fixing.

File: tiatoolbox/visualization/tileserver.py
Comment (John-P) [inline tiatoolbox/visualization/tileserver.py:18]: Might need to add to the docstring here to explain new behaviour, e.g. how layer names are determined and coloursiation etc.

**Review by John-P — CHANGES_REQUESTED** [review]

**Comment by John-P** [comment]
@measty Can you please mark the comments as resolved if they have been addressed.

**Pull request merged**
